lib/SILGen/SILGenBuiltin.cpp

```diff
@@ -1042,28 +1042,60 @@ static ManagedValue emitBuiltinTypeTrait(SILGenFunction &SGF,
 }
 
 // SWIFT_ENABLE_TENSORFLOW
-static ManagedValue emitBuiltinAutoDiffGetJVP(SILGenFunction &SGF,
-                                              SILLocation loc,
-                                              SubstitutionMap substitutions,
-                                              Expr *argument,
-                                              SGFContext C) {
-  auto argVal = SGF.emitRValue(argument);
-  auto jvp = SGF.getBuilder().createAutoDiffFunctionExtract(
-      loc, AutoDiffFunctionExtractee::JVP, /*differentiationOrder*/ 1,
-      std::move(argVal).forwardAsSingleValue(SGF, loc));
-  return ManagedValue::forUnmanaged(jvp);
+static ManagedValue emitBuiltinAutoDiffApplyAssociatedFunction(
+    AutoDiffAssociatedFunctionKind kind, SILGenFunction &SGF, SILLocation loc,
+    SubstitutionMap substitutions, Expr *argument, SGFContext C) {
+  auto *argTuple = cast<TupleExpr>(argument);
+  assert(argTuple->getNumElements() == 2);
+  auto origFnVal =
+      SGF.emitRValueAsSingleValue(argTuple->getElement(0)).getUnmanagedValue();
+  auto origFnArgVal =
+      SGF.emitRValueAsSingleValue(argTuple->getElement(1)).forward(SGF);
+  auto assocFn = SGF.B.createAutoDiffFunctionExtract(
+          loc, kind, /*differentiationOrder*/ 1, origFnVal);
+  auto assocFnType = assocFn->getType().getAs<SILFunctionType>();
+  assert(assocFnType->getNumResults() == 2);
+  // Emit an l-value when the function has indirect results.
+  if (assocFnType->hasIndirectFormalResults()) {
+    assert(origFnArgVal->getType().isAddress());
+    auto indResBuffer =
+        SGF.getBufferForExprResult(loc, assocFnType->getAllResultsType(), C);
+    SILValue applyArgs[2] {
+      SGF.B.createTupleElementAddr(loc, indResBuffer, 0),
+      origFnArgVal
+    };
+    auto differential = SGF.B.createApply(loc, assocFn, applyArgs,
+                                          /*isNonThrowing*/ false);
+    SGF.B.createStore(loc, differential,
+                      SGF.B.createTupleElementAddr(loc, indResBuffer, 1),
+                      StoreOwnershipQualifier::Init);
+    return ManagedValue::forLValue(indResBuffer);
+  }
+  // Emit an r-value tuple otherwise.
+  assert(origFnArgVal->getType().isObject());
+  auto resultTuple =
+      SGF.B.createApply(loc, assocFn, {origFnArgVal}, /*isNonThrowing*/ false);
+  return ManagedValue::forUnmanaged(resultTuple);
```

Comment:
should this be `ManagedValue::forOwnedObjectRValue`?

Reply:
`ManagedValue::forOwnedObjectRValue` takes a clean-up. `SGF.emitManagedRValueWithCleanup` seems to be the right one to use.